modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TabPaneSkinHeaderOrderTest.java

```diff
@@ -0,0 +1,289 @@
+    }
+
+    @Test
+    public void testAddSingleAtBeginning() {
+        tabPane.getTabs().add(0, new Tab("t0"));
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testAddSingleAtEnd() {
+        tabPane.getTabs().add(new Tab("t6"));
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testAddSingleAtEndIndex() {
+        tabPane.getTabs().add(tabPane.getTabs().size(), new Tab("t6"));
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testAddSingleInMiddle() {
+        tabPane.getTabs().add(2, new Tab("tm"));
+        assertSyncTabHeaders();
+    }
+
+    @Ignore("JDK-8245528")
+    @Test
+    public void testAddSingleExistingTab() {
+        tabPane.getTabs().add(1, tabPane.getTabs().get(3));
+        assertSyncTabHeaders();
+    }
+
+    @Ignore("JDK-8245528")
+    @Test
+    public void testAddMultipleExistingTabsAtBeginning() {
+        List<Tab> added = List.of(tabPane.getTabs().get(3),
+                                  tabPane.getTabs().get(4));
+        tabPane.getTabs().addAll(0, added);
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testAddCollectionAtBeginning() {
+        List<Tab> added = List.of(new Tab("t-1"), new Tab("t0"));
+        tabPane.getTabs().addAll(0, added);
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testAddMultipleAtEnd() {
+        tabPane.getTabs().addAll(new Tab("t6"), new Tab("t7"));
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testAddCollectionAtEnd() {
+        List<Tab> added = List.of(new Tab("t6"), new Tab("t7"));
+        tabPane.getTabs().addAll(tabPane.getTabs().size(), added);
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testAddCollectionInMiddle() {
+        List<Tab> added = List.of(new Tab("tm1"), new Tab("tm2"));
+        tabPane.getTabs().addAll(2, added);
+        assertSyncTabHeaders();
+    }
+
+    @Test
+    public void testInitialTabOrder() {
+        assertSyncTabHeaders();
+    }
+
+    protected void assertSyncTabHeaders() {
+        assertSyncHeaders(tabPane.getTabs(),
+                          TabPaneSkinShim.getTabHeaders(tabPane));
+    }
+
+    protected void assertSyncHeaders(List<Tab> tabs, List<Node> headers) {
+        assertEquals("sanity: same size", tabs.size(), headers.size());
+        for (int i = 0; i < tabs.size(); i++) {
+            Tab headerTab = (Tab) headers.get(i).getProperties().get(Tab.class);
+            assertSame("tab at " + i + ", is: " + tabs.get(i).getText()
+                + " but in header it is: " + headerTab.getText(),
+                tabs.get(i), headerTab);
+        }
+    }
+
+
+//----------------- setup and initial
+    @Test
+    public void testSetupState() {
+        assertNotNull(tabPane);
+        List<Node> expected = List.of(tabPane);
+        assertEquals(expected, root.getChildren());
+    }
+
+    @After
+    public void cleanup() {
+        stage.hide();
```

Comment:
I recommend to restore animation here.

Reply:
Never mind. This is not needed since the test constructs a new TabPane instance for each test in the `@Before` method.